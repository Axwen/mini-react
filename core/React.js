
function createTextNode(text) {
    return {
        type: 'TEXT_ELEMENT',
        props: {
            nodeValue: text,
            children: [],
        }
    }
}


function createElement(type, props, ...children) {
    return {
        type,
        props: {
            ...props,
            children: children.map(child => {
                return typeof child === 'string' ? createTextNode(child) : child
            })
        }
    }
}

function render(el, container) {
    const { type, props: { children = [], ...props }, } = el
    const dom = type === 'TEXT_ELEMENT' ? document.createTextNode('') : document.createElement(el.type)

    Object.keys(props).forEach(key => {
        dom[key] = props[key]
    })

    children.forEach(child => {
        render(child, dom)
    })
    container.appendChild(dom)
}


const React = {
    createElement,
    render
}

export default React;