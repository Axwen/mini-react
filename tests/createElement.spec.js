import React from '../core/React.js';

import { describe, it, expect } from 'vitest'


describe('createElement', () => {
    it('should create a div element', () => {
        const element = React.createElement('div', null, 'Hello, world!');
        expect(element).toMatchSnapshot(`
            {
              "props": {
                "children": [
                    {
                        "props":{
                            "children":[],
                            "nodeValue": "Hello, world!",
                        },
                        "type": "TEXT_ELEMENT",
                    }
                ],
              },
              "type": "div",
            }
        `)
    })
})