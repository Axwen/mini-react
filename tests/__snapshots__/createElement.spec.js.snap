// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`createElement > should create a div element > 
            {
              "props": {
                "children": [
                    {
                        "props":{
                            "children":[],
                            "nodeValue": "Hello, world!",
                        },
                        "type": "TEXT_ELEMENT",
                    }
                ],
              },
              "type": "div",
            }
         1`] = `
{
  "props": {
    "children": [
      {
        "props": {
          "children": [],
          "nodeValue": "Hello, world!",
        },
        "type": "TEXT_ELEMENT",
      },
    ],
  },
  "type": "div",
}
`;
